File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    static int N;
    static int[] dp = new int[300];
    static int[] stair = new int[300];
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            stair[i] = Integer.parseInt(br.readLine());
        }
        dp[0] = stair[0];
        dp[1] = Math.max(stair[0]+stair[1], stair[1]);
        dp[2] = Math.max(stair[0]+stair[2],stair[1]+stair[2]);
        for (int i = 3; i < N; i++) {
            dp[i] = Math.max(dp[i-2] + stair[i], dp[i-3] + stair[i-1] + stair[i]);
        }
        System.out.println(dp[N-1]);
    }
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys
N = int(sys.stdin.readline())
stairs = []
dp = []
for _ in range(N):
    stairs.append(int(sys.stdin.readline()))

dp.append(stairs[0])
if N >= 2:
    dp.append(stairs[0] + stairs[1])
if N >= 3:
    dp.append(max(stairs[1]+stairs[2],stairs[0]+stairs[2]))

for i in range(3, N):
    dp.append(max(dp[i-3] + stairs[i-1] + stairs[i],
                  dp[i-2] + stairs[i]))

print(dp[-1])